prompt-service/src/unstract/prompt_service/main.py

```diff
@@ -693,29 +693,51 @@ def run_retrieval(  # type:ignore
     retrieval_type: str,
     metadata: dict[str, Any],
 ) -> tuple[str, str]:
+    context: str = ""
     prompt = output[PSKeys.PROMPTX]
     if retrieval_type == PSKeys.SUBQUESTION:
-        subq_prompt = (
-            f"Generate a sub-question from the following verbose prompt that will"
-            f" help extract relevant documents from a vector store:\n\n{prompt}"
+        subq_prompt: str = (
+            f"I am sending you a verbose prompt \n \n Prompt : {prompt} \n \n"
+            "Generate set of specific subquestions "
+            "from the prompt which can be used to retrive "
+            "relevant context from vector db. "
+            "Ensure that each subquestion is distinct and relevant"
+            "to the different facets of the original query. Do not"
+            "add subquestions for details not mentioned in the original "
+            "prompt. The goal is to maximize retrieval accuracy"
+            " using these subquestions. Please note that, there are cases where the "
+            "response might have a list of answers. The subquestions must not miss out "
+            "any values in these cases. "
+            "Output should be a list of comma seperated "
+            "subquestion prompts. Do not change this format. \n \n "
+            " Subquestions : "
         )
-        prompt = run_completion(
+        subquestions = run_completion(
             llm=llm,
             prompt=subq_prompt,
         )
-    context = _retrieve_context(output, doc_id, vector_index, prompt)
-
-    if not context:
-        # UN-1288 For Pinecone, we are seeing an inconsistent case where
-        # query with doc_id fails even though indexing just happened.
-        # This causes the following retrieve to return no text.
-        # To rule out any lag on the Pinecone vector DB write,
-        # the following sleep is added
-        # Note: This will not fix the issue. Since this issue is inconsistent
-        # and not reproducible easily, this is just a safety net.
-        time.sleep(2)
+        subquestion_list = subquestions.split(",")
+        for each_subq in subquestion_list:
+            retrived_context = _retrieve_context(
+                output, doc_id, vector_index, each_subq
+            )
+            context = "".join([context, retrived_context])
+
+    if retrieval_type == PSKeys.SIMPLE:
+
         context = _retrieve_context(output, doc_id, vector_index, prompt)
 
+        if not context:
```

Comment:
@harini-venkataraman Shouldn't this check happen after retrieving sub-question context above too?

Reply:
@gaya3-zipstack Can you please confirm if the inconsistent issue related to pinecone  still exists? 
If so, the sub question retrieval fetches context for every iteration, let me if this check is needed here?

Reply:
As per my knowledge, nothing might have changed w.r.t  Pinecone there @harini. But is that what @hari-kuriakose is asking about? I am not clear.

Reply:
@hari-kuriakose As per discussion with @gaya3-zipstack, a pointers.
1. The sleep added here is only a potential fix for inconsistency in Pinecone serverless retrieval. We cannot afford to add this sleep in a loop, expecting a infinite loop.
2. Also, since the sub questions are returned from LLM, we will never have a way to differentiate genuine cases of empty context. 

Considering these challenges and niche usecase, we can leave it this way. Added a comment in code for good developer experience.

Reply:
@harini-venkataraman One thing though, the loop will be always finite. We will do this check only once at max for each sub-question as per existing logic.

Reply:
@hari-kuriakose Thanks for catching that.